wpimath/src/main/native/cpp/jni/WPIMathJNI.cpp

```diff
@@ -187,6 +188,76 @@ Java_edu_wpi_first_math_WPIMathJNI_pow
   env->SetDoubleArrayRegion(dst, 0, rows * rows, Apow.data());
 }
 
+/*
+ * Class:     edu_wpi_first_math_WPIMathJNI
+ * Method:    expPose3d
+ * Signature: (DDDDDDDDDDDDD)[D
+ */
+JNIEXPORT jdoubleArray JNICALL
+Java_edu_wpi_first_math_WPIMathJNI_expPose3d
+  (JNIEnv* env, jclass, jdouble poseX, jdouble poseY, jdouble poseZ,
+   jdouble poseQw, jdouble poseQx, jdouble poseQy, jdouble poseQz,
+   jdouble twistDx, jdouble twistDy, jdouble twistDz, jdouble twistRx,
+   jdouble twistRy, jdouble twistRz)
+{
+  frc::Pose3d pose{
+      units::meter_t{poseX}, units::meter_t{poseY}, units::meter_t{poseZ},
+      frc::Rotation3d{frc::Quaternion{poseQw, poseQx, poseQy, poseQz}}};
+  frc::Twist3d twist{units::meter_t{twistDx},  units::meter_t{twistDy},
+                     units::meter_t{twistDz},  units::radian_t{twistRx},
+                     units::radian_t{twistRy}, units::radian_t{twistRz}};
+
+  frc::Pose3d result = pose.Exp(twist);
+
+  jdoubleArray resultArray = env->NewDoubleArray(7);
+  jdouble* nativeResultArray =
+      env->GetDoubleArrayElements(resultArray, nullptr);
+  const auto& resultQuaternion = result.Rotation().GetQuaternion();
+  nativeResultArray[0] = result.X().value();
+  nativeResultArray[1] = result.Y().value();
+  nativeResultArray[2] = result.Z().value();
+  nativeResultArray[3] = resultQuaternion.W();
+  nativeResultArray[4] = resultQuaternion.X();
+  nativeResultArray[5] = resultQuaternion.Y();
+  nativeResultArray[6] = resultQuaternion.Z();
+  env->ReleaseDoubleArrayElements(resultArray, nativeResultArray, 0);
+  return resultArray;
+}
+
+/*
+ * Class:     edu_wpi_first_math_WPIMathJNI
+ * Method:    logPose3d
+ * Signature: (DDDDDDDDDDDDDD)[D
+ */
+JNIEXPORT jdoubleArray JNICALL
+Java_edu_wpi_first_math_WPIMathJNI_logPose3d
+  (JNIEnv* env, jclass, jdouble startX, jdouble startY, jdouble startZ,
+   jdouble startQw, jdouble startQx, jdouble startQy, jdouble startQz,
+   jdouble endX, jdouble endY, jdouble endZ, jdouble endQw, jdouble endQx,
+   jdouble endQy, jdouble endQz)
+{
+  frc::Pose3d startPose{
+      units::meter_t{startX}, units::meter_t{startY}, units::meter_t{startZ},
+      frc::Rotation3d{frc::Quaternion{startQw, startQx, startQy, startQz}}};
+  frc::Pose3d endPose{
+      units::meter_t{endX}, units::meter_t{endY}, units::meter_t{endZ},
+      frc::Rotation3d{frc::Quaternion{endQw, endQx, endQy, endQz}}};
+
+  frc::Twist3d result = startPose.Log(endPose);
+
+  jdoubleArray resultArray = env->NewDoubleArray(6);
+  jdouble* nativeResultArray =
+      env->GetDoubleArrayElements(resultArray, nullptr);
+  nativeResultArray[0] = result.dx.value();
+  nativeResultArray[1] = result.dy.value();
+  nativeResultArray[2] = result.dz.value();
+  nativeResultArray[3] = result.rx.value();
+  nativeResultArray[4] = result.ry.value();
+  nativeResultArray[5] = result.rz.value();
+  env->ReleaseDoubleArrayElements(resultArray, nativeResultArray, 0);
```

Comment:
This can be simplified to `return wpi::MakeJDoubleArray(env, {{result.dx.value(), ...}});`  (see https://github.com/wpilibsuite/allwpilib/blob/main/wpiutil/src/main/native/include/wpi/jni_util.h#L551)